azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,377 @@
+    def _create_base_dirs():
+        LogCollector._mkdir(_LOG_COLLECTOR_DIR)
+        LogCollector._mkdir(_TRUNCATED_FILES_DIR)
+
+    @staticmethod
+    def _set_logger():
+        _f_handler = logging.FileHandler(_OUTPUT_RESULTS_FILE_PATH, encoding="utf-8")
+        _f_format = logging.Formatter(fmt='%(asctime)s %(levelname)s %(message)s',
+                                      datefmt=u'%Y-%m-%dT%H:%M:%SZ')
+        _f_format.converter = time.gmtime
+        _f_handler.setFormatter(_f_format)
+        _LOGGER.addHandler(_f_handler)
+        _LOGGER.setLevel(logging.INFO)
+
+    @staticmethod
+    def _run_shell_command(command, stdout=subprocess.PIPE, log_output=False):
+        """
+        Runs a shell command in a subprocess, logs any errors to the log file, enables changing the stdout stream,
+        and logs the output of the command to the log file if indicated by the `log_output` parameter.
+        :param command: Shell command to run
+        :param stdout: Where to write the output of the command
+        :param log_output: If true, log the command output to the log file
+        """
+        def format_command(cmd):
+            return " ".join(cmd) if isinstance(cmd, list) else command
+
+        def _encode_command_output(output):
+            return ustr(output, encoding="utf-8", errors="backslashreplace")
+
+        try:
+            process = subprocess.Popen(command, stdin=subprocess.PIPE, stdout=stdout, stderr=subprocess.PIPE, shell=False)
+            stdout, stderr = process.communicate()
+            return_code = process.returncode
+        except Exception as e:
+            error_msg = u"Command [{0}] raised unexpected exception: [{1}]".format(format_command(command), ustr(e))
+            _LOGGER.error(error_msg)
+            return
+
+        if return_code != 0:
+            encoded_stdout = _encode_command_output(stdout)
+            encoded_stderr = _encode_command_output(stderr)
+            error_msg = "Command: [{0}], return code: [{1}], stdout: [{2}] stderr: [{3}]".format(format_command(command),
+                                                                                                 return_code,
+                                                                                                 encoded_stdout,
+                                                                                                 encoded_stderr)
+            _LOGGER.error(error_msg)
+            return
+
+        if log_output:
+            msg = "Output of command [{0}]:\n{1}".format(format_command(command), _encode_command_output(stdout))
+            _LOGGER.info(msg)
+
+    @staticmethod
+    def _expand_must_collect_files():
+        # Match the regexes from the MUST_COLLECT_FILES list to existing file paths on disk.
+        manifest = []
+        for path in _MUST_COLLECT_FILES:
+            manifest.extend(sorted(glob.glob(path)))
+
+        return manifest
+
+    @staticmethod
+    def _parametrize_filepath(path):
```

Comment:
Why is this function needed now since you've already parameterized it earlier?

Reply:
+1

Reply:
Not needed, good catch. Forgot to remove it.